### DISCUSSION TITLE
[Issue] Unable to signup to newsletter only

### DISCUSSION THREAD

**Discussion by danf97** [Shopify/storefront-api-feedback#124] [Help]
Hi all,

I’m working on a project based on hydrogen and I’m using the GraphQL Storefront API.
The GraphQL documentation allows us to create or update a customer but there is no documentation if the user only wants to subscribe the newsletter.
There is no mutation to do it? Does it exist but it's not documented? 

[GraphQL Storefront API Documentation](https://shopify.dev/api/storefront/2022-04/mutations/customerCreate
)

Many thanks,
Daniel

**Comment by danf97** [comment]
Hi @lordofthecactus ,

Is this the best way to report this issue? It's affecting the Hydrogen framework.

Cheers,
Daniel

**Reply by lordofthecactus** [reply]
Yeah. This is alright for giving feedback about the storefront API 💯

**Comment by lordofthecactus** [comment]
I also could not find if it is possible to do only with storefront API 🤔

**Not sure if this will work or if this is the best way, I'll ask my colleagues to confirm:** 
Seems you can create a customer (without password) with the Admin API and enable marketing:
https://shopify.dev/api/admin-graphql/2022-07/mutations/customerCreate

Here is an example of how to use the Admin API with Hydrogen
https://github.com/Shopify/hydrogen/pull/1788

**Reply by danf97** [reply]
Please let me know if this feature will be soon available on the Storefront API.

Thank you for your help!!

**Reply by danf97** [reply]
Hi @lordofthecactus 

I've implemented this but the customer doesn't get any notification/email to confirm the subscription. Thats not very gdpr complaint.
Any solution for this? Is it possible to trigger a notification/email with the Admin Api? Only the Storefront API is triggering notifications automatically...

All the best,
Daniel

**Reply by benjaminsehl** [reply]
You could create your own email automation using Shopify Flow (or a third party like Klaviyo)

When you do that mutation it does set the consent level at single opt in.

I would not recommend using Admin API for this though, as it's not really meant for that use case. We're working actively on solving for newsletters with the Storefront API, hope to have something soon. 

In the meantime, you _could_ build something with Autocode.com or Gadget.dev that could handle it for you.

**Reply by danf97** [reply]
Hi @benjaminsehl,

Thank's for your help on this! 
I'll consider those options :) 

And nice work on the Hydrogen framework, it's super cool!

All the best,
Daniel

**Comment by bencarrr** [comment]
@benjaminsehl Saw this is being actively worked on since August - any update on the handling of newsletters in the Storefront API?

**Reply by ryanleichty** [reply]
Coming here for an answer as well 👋🏻

**Reply by kristofferso** [reply]
Same here 😎

**Reply by KyleWiteck** [reply]
I agree, very normal functionality for all types of websites.

**Comment by danf97** [comment]
Hello @lordofthecactus and @benjaminsehl,

Sorry to pin you. Hope your are doing well. 
After a year on this issue is there any news on it?

Best,
Daniel

**Reply by benjaminsehl** [reply]
Hey Daniel -- there's been a larger overall change in how customers are handled in the API that's been blocking this.

Good news is we're now getting close to that change going live, and will work to accommodate this use case soon after.

I don't have a date that I can quote in terms of when to expect it, but can assure you this is at the very top of my list in terms of things I'm pushing on for API capabilities.

Thanks for your patience!

**Reply by ryanleichty** [reply]
Thanks for the update @benjaminsehl! Been waiting for this one for a while too

**Reply by benjaminsehl** [reply]
Totally get it! I have 2 headless stores myself that I want to do this on.

Feel free to keep leaving feedback on other gaps you want closed and tag me in it.

**Reply by danf97** [reply]
Thanks for the update @benjaminsehl!

**Reply by kristofferso** [reply]
Any news about this @benjaminsehl? I read the storefront api got a major upgrade this summer but haven't been able to find out if any of the updates will cover newsletter subscription.

**Reply by Aneeq-Tahir** [reply]
Any updates? @benjaminsehl

**Reply by benjaminsehl** [reply]
Fielding this to @nickpinto, who's team works on the upcoming Customer Accounts API, which I believe will be the right home for this.

**Reply by nickpinto** [reply]
Hey folks! Sorry, a little late to the game. The [Customer Account API](https://shopify.dev/docs/custom-storefronts/building-with-the-customer-account-api/) supports [email marketing subscribe](https://shopify.dev/docs/api/customer/2024-01/mutations/customerEmailMarketingSubscribe) and [unscubscribe](https://shopify.dev/docs/api/customer/2024-01/mutations/customerEmailMarketingUnsubscribe). Does that solve what you're looking for?

**Reply by ryanleichty** [reply]
@nickpinto Awesome, thanks for the update! I know there's a handful of threads in Discord that have been waiting/looking for answers to this as well.

Does `customerEmailMarketingSubscribe` automate a confirmation email? I know that was an issue with the workaround when using the admin API to subscribe customers to email marketing.

**Reply by nickpinto** [reply]
Hey @ryanleichty 👋🏽. These mutations don't trigger an email confirmation to the buyer. The situation is slightly different with the Customer Account API when compared to the Admin API because the API requires an access token that represents a logged-in and authenticated buyer to be able to use any of the mutations. Logging-in serves as a verification process to confirm that the customer owns that email address and deliberately wants to subscribe or unsubscribe from the email marketing. 

Do you think there are still scenarios that would require an email also? If you still want to send an email out, you could create your own email automation using Shopify Flow

**Reply by Alexanderdomino** [reply]
Hi @nickpinto ✌️ 

Sorry to pop in!

These are definitely cool additions already, but I really think it would be beneficial with an option to subscribe to a newsletter without being logged in and possibly with email confirmation as this is the normal behavior on most webshops.

Do you think this will be coming to the Customer API as well?

**Reply by ryanleichty** [reply]
Oh yikes, yeah we definitely need the option to subscribe to email marketing without being logged in / a registered customer. I feel like that has to be the most common email subscription scenario.

**Reply by lucasprins** [reply]
Is this not already possible with a headless storefront? Beause IIRC the demo store has a newsletter field in the footer that works without registered customers. I just couldn't find the equivalent mutation for the storefront api.

**Reply by Alexanderdomino** [reply]
@lucasprins no newsletterform on https://hydrogen.shop

**Reply by jonrrivera** [reply]
I just spent 6 hours figuring this out, just to get to this thread. I'm out

**Reply by lucasprins** [reply]
@Alexanderdomino oh yeah sorry I meant the "Online Store" with the default theme of Shopify.

**Reply by ryanleichty** [reply]
I remember trying to find a solution to this over a year ago for a project I haven't touched in a while. Still no solid answers yet. The communication from Shopify isn't as transparent as I would like.

**Reply by nickpinto** [reply]
@Alexanderdomino and @ryanleichty, yes, we are looking at a solution for subscribing without being authenticated. Unfortunately, I don't have an ETA to share at this time.

**Reply by jonrrivera** [reply]
so there is no way to have a simple email submission on shopify hydrogen? would you recommend looking into leveraging klaviyo or another similar dependency ? @nickpinto

**Reply by nickpinto** [reply]
Not at this moment @jonrrivera. For now, leveraging Klaviyo or another 3P partner is the only thing I can suggest. Completely aligned that this is needed and I understand the priority.

**Reply by kellymilligan** [reply]
Hey @nickpinto, has there been progress made on this request since this previous update? Thanks!

**Comment by Hooman-studio** [comment]
Still looking for this as well, to clarify like others did. Newsletter signup without customer signup. Like basically any newsletter signup form on any website, and also is supported on online stores. 

This thread seems to imply work on this started in Summer 2022, now almost 2 years ago, is there any update for when this will be available?

**Reply by nickpinto** [reply]
Hey @Hooman-studio 👋🏽. I'm definitely aligned on the need for this and we are looking at a solution for newsletter sign-up without customer login. oof - can see how this thread implied that work started last year, it hasn't, as there were a few major blockers to us making progress on it. There isn't a timeline for this to release, but I want to have this out as soon as we can. In the meantime, if leveraging a 3P like Klaviyo can work, I would recommend exploring that. 

Thanks for your patience!

**Reply by Hooman-studio** [reply]
Hi Nick, thank you for your response! Really appreciate it. For other people looking I got the subscribe to work with the admin API. Here is my nodejs implementation (within nextjs):

```typescript
export const customerCreateMutation = /* GraphQL */ `
  mutation customerCreate($input: CustomerInput!) {
    customerCreate(input: $input) {
      customer {
        id
        email
      }
      userErrors {
        field
        message
      }
    }
  }
`;

export const customerUpdateMarketingMutation = /* GraphQL */ `
  mutation customerEmailMarketingConsentUpdate($input: CustomerEmailMarketingConsentUpdateInput!) {
    customerEmailMarketingConsentUpdate(input: $input) {
      customer {
        id
      }
      userErrors {
        field
        message
      }
    }
  }
`;
export const customerCreateForMarketingEmail = async (input: { email: string }) => {
  const res = await shopifyAdminFetch<{
    data: {
      customerCreate: {
        customer: {
          id: string;
          email: string;
        };
      };
    };
    variables: {
      input: {
        email: string;
      };
    };
  }>({
    cache: 'no-store',
    query: customerCreateMutation,
    variables: {
      input: {
        email: input.email
      }
    }
  });

  const updateRes = await shopifyAdminFetch<{
    variables: {
      input: {
        customerId: string;
        emailMarketingConsent: {
          marketingOptInLevel: 'CONFIRMED_OPT_IN' | 'UNKNOWN' | 'SINGLE_OPT_IN';
          marketingState: 'SUBSCRIBED' | 'PENDING' | 'UNSUBSCRIBED';
        };
      };
    };
  }>({
    cache: 'no-store',
    query: customerUpdateMarketingMutation,
    variables: {
      input: {
        customerId: res.data.customerCreate.customer.id,
        emailMarketingConsent: {
          marketingState: 'SUBSCRIBED',
          marketingOptInLevel: 'SINGLE_OPT_IN'
        }
      }
    }
  });
  console.log(updateRes);
  return res.data.customerCreate.customer;
};
```

It also works if you just do the createCustomer mutation without the update and set the `emailMarketingConsent` directly.

Now this works, but for some reason (even though I do the subscription update separately) any email automation that should trigger when someone subscribes is not working. If that is somehow possible then all my needs are met with the admin api as well.

**Reply by sampsondaniel** [reply]
Hello Hooman ! Do you have a component that uses those mutations for the newsletter signup ?

Thanks for your solution !

**Reply by nhm1990** [reply]
That's pretty frustrating. I've been sitting on this problem for half the week and have only just come across this post...

Even with Flow I have not yet managed to trigger such a mail. 

You can say: 
If CustomerCreate then .... 

But the email marketing mail only goes out IF the user has already done double opt-in.

**Comment by martin-lacko-rebood** [comment]
Any update about this by any chance? We are attempting to migrate from liquid themes to hydrogen. While [shopify themes](https://shopify.dev/docs/api/liquid/tags/form#form-customer) seem to offer this functionality out of the box it appears the storefront API (and if I understand correctly admin API as well) is still lacking behind. It's a shame as we would like to stick as close to native shopify as possible and relay on as few third party external tools as possible. Some update on this would be great. Thank you!

**Reply by nhm1990** [reply]
For me, Klaviyo has so far proved to be a great alternative. The app is much better developed and offers more functionality and configuration options. Everything also stays in sync.

**Comment by rafaelcaceres** [comment]
We have this now, is it new?

https://shopify.dev/docs/api/storefront/2025-01/mutations/customerCreate

**Reply by joshistoast** [reply]
Not new. As you can see on the input type, a password is still required.

**Reply by rafaelcaceres** [reply]
Yeah, my bad! Thanks

**Comment by yfi** [comment]
Very frustrating. I thought i could emulate the email with a flow but if I create the user with marketingState Pending the flow will refuse to send a marketing email because the user is not subscribed. 🤦‍♂️

my current working plan:

~~1. Create an optin_url_metafield on customer
2. HEADLESS STOREFRONT -> admin api -> create a customer with:
marketingState: CONFIRMED, tags: [newsletter-signup], optin_url_metafiled = link to a dynamic route with some random hash: ie /api/opt-in/{uuid}
3. FLOW: on customer create -> check for newsletter-signup tag and the optin_url_metafield  >then> send simulated marketing email with optin_url  >then> send admin api request to downgrade marketingState to PENDING
4. User receives email, clicks link, that takes you to yourheadless.store.com/optin_url
5. HEADLESS STOREFRONT optin_url route calls shopify admin api to again set the marketingState to SUBSCRIBED, remove optin_url_metafiled, remove the tag
6. redirect to a welcome page~~

[edit: Updated method using tags only. Much simpler. See below]

😅

It's late but i'll test this monstrosity tomorrow

**Reply by yfi** [reply]
for existing customers using a newsletter form.... there's a customer tag created / removed trigger that could work...

**Reply by yfi** [reply]
It Works. But it's much simpler to leverage the customer tag system than rely on metafields.
Here's the simplified setup:

### 1. Signup Form upserts user with SUBSCRIBED marketingConsentState
```gql
mutation customerSet($input: CustomerSetInput!, $identifier: CustomerSetIdentifiers) {
  customerSet(input: $input, identifier: $identifier) {
     customer {
            id
            email
          }
          userErrors {
            field
            message
          }
       }
    }


input{
  email
      emailMarketingConsent: {
        marketingOptInLevel: 'CONFIRMED_OPT_IN',
        marketingState: 'SUBSCRIBED'
      }
}


```
then save a unique opt in link as a tag... 

```gql
   
mutation tagsAdd($id: ID!, $tags: [String!]!) {
        tagsAdd(id: $id, tags: $tags) {
          node {
            id
            ... on Customer {
              tags
            }
          }
          userErrors {
            field
            message
          }
        }
      }

variables {
    id
    tags:[`opt-in-${sha256(customer.email}`]
}

```
You only want one tag with an opt-in- prefix per customer for this to work.
I'm using a simple sha256 hash so this tag will always be the same and generate an identical url everytime the form is sent.
If you want unique urls per signup attempt (ie. user subscribed, then unsubscribed, and is subscribing again ) you'll need to handle that logic and replace existing opt-in-{hash} tags with the fresh one.

### 2. Shopify Flow. Triggered when opt-in-* tag is added, sends confirmation email using the email marketing app and immediately downgrades customer marketingState to PENDING

```
1. trigger: customer tag added
2. condition: if tag contains opt-in
3. send email: with a link like <a href="{%- for tag in customer.tags -%}{%- if tag contains 'opt-in-' -%}https://your-headless-store.com/customer/{{ tag }}{% break %}{% endif %}{% endfor %}">Confirm Subscription</a>
5. make admin api request: customerEmailMarketingConsentUpdate {
  "input": {
    "customerId": "{{customer.id}}",
    "emailMarketingConsent": {
      "marketingOptInLevel": "CONFIRMED_OPT_IN",
      "marketingState": "PENDING"
    }
  }
}
```
### 3. Headless opt-in route... 

my route is same as the tag I set: `/customer/opt-in-[hash]`
First, find Customer by tag

```gql
customers(first:1, query: "tag:opt-in-${hash} AND accepts_marketing:false"){
  ...
}
```

if customer is found, update email consent

```gql
mutation customerEmailMarketingConsentUpdate($input: CustomerEmailMarketingConsentUpdateInput!) {
  customerEmailMarketingConsentUpdate(input: $input) {
    customer {
      id
      email
      emailMarketingConsent {
        marketingState
        marketingOptInLevel
      }
    }
    userErrors {
      field
      message
    }
  }
}


 input {
    customerId: customer.id,
        emailMarketingConsent: {
            marketingOptInLevel: "CONFIRMED_OPT_IN",
            marketingState: "SUBSCRIBED"
         }
}

```

then clean up opt-in tag and redirect to a success page.

you can set the tracking options in the email marketing app to "do not track" for this to be gdpr coompliant.

**Reply by ryanleichty** [reply]
The fact that this is so difficult to begin with 🤦🏼‍♂️

**Comment by williamkhshea** [comment]
Any chance this will get fixed this year?

**Reply by williamkhshea** [reply]
@benjaminsehl is there any update on this issue?
